Allow dragging multiple nodes to filesystem and create scenes without dialog and using {nodename}.tscn

Xenthera opened the pull request:
<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

I grew tired of dragging each node (3d model in this case) inside a gltf file into the file system one by one to create new scenes. This small change allows you to drag multiple nodes into the filesystem and create .tscn files, bypassing the save file dialogue and just saving the new scenes with the same name the node had. This only works with dragging, and not rightclicking multiple nodes. Dragging multiple nodes into the filesystem will obviously save to the folder the drag ended on.

Commit:
Added ability to create scenes from multiple selected nodes instead of just one.

Commit:
Removed garbage

KoBeWi commented:
Dropping nodes in empty area results in generic error:

https://github.com/godotengine/godot/assets/2223172/66c08ea1-b15b-4cf4-b024-b51e3e15a07b

It should save them in `res://`. And also the error should provide more details.

Creating multiple scenes will create undo action for each one separately. It should be a single action.

Commit:
PR Feedback

Fixed bug when dragging multiple nodes to empty area
Code refactor for duplication removal

Commit:
Clang-Format

Xenthera commented:
> Dropping nodes in empty area results in generic error:
> 
>  VqKyGkDJ9q.mp4 
> It should save them in `res://`. And also the error should provide more details.
> 
> Creating multiple scenes will create undo action for each one separately. It should be a single action.

Thanks for the feedback, bug is fixed when dragging to blank area. It behaved differently on macOS and actually succeeded but I added an explicit check. As far as undo goes, the undo code looks like magic to me so it'll be a minute before I get those changes implemented.

KoBeWi commented:
The easiest way is to make use of nested actions. If you create an action, any other action created before the parent is commited will be merged into single action.

Xenthera committed:
clang-format changed this, static analysis did not like it.

Xenthera committed:
Undo Support

Previously you would have to undo n time for n scenes. Now dragging multiple nodes is treated as one undo event and will revert all scenes previously saved.

Xenthera commented:
> The easiest way is to make use of nested actions. If you create an action, any other action created before the parent is commited will be merged into single action.

Thanks, I would've have spun my wheels without that bit of knowledge.

Xenthera committed:
Skip if file exists as to prevent unintentional file overwriting.

If a node already exists as a scene in a multidrag operation, that will be skipped. This prevents unwanted overwriting as it wasn't obvious the file had been overwritten before.

Xenthera committed:
Small bug fix

Would not allow file dialogue to pop up if dragging single file. Unintended behaviour fixed.

KoBeWi commented:
Dropping nodes with invalid filenames will result in short editor freeze and error spam:

https://github.com/godotengine/godot/assets/2223172/f5445b0e-3a06-4de2-a6aa-3746b6795101

You should use `validate_filename()` when saving file.

KoBeWi commented on editor/scene_tree_dock.h:
```diff
@@ -33,6 +33,7 @@
 
 #include "editor/gui/scene_tree_editor.h"
 #include "editor/script_create_dialog.h"
+#include "editor_undo_redo_manager.h"
```
This should be in cpp file.

KoBeWi commented on editor/scene_tree_dock.h:
```diff
@@ -161,6 +162,7 @@ class SceneTreeDock : public VBoxContainer {
 	ReparentDialog *reparent_dialog = nullptr;
 	EditorQuickOpen *quick_open = nullptr;
 	EditorFileDialog *new_scene_from_dialog = nullptr;
+	String current_dir;
```
You could reuse current directory from `new_scene_from_dialog`.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -974,60 +972,69 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
+					new_scene_from_dialog->set_title(TTR("Save New Scene As..."));
+					new_scene_from_dialog->popup_file_dialog();
+				} else {
+					if (da->file_exists(current_dir + existing)) {
```
```suggestion
					if (da->file_exists(current_dir.path_join(existing))) {
```
Same below.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -974,60 +972,69 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
+				if (tocopy != editor_data->get_edited_scene_root() && !tocopy->get_scene_file_path().is_empty()) {
+					accept->set_text(TTR("Can't save the branch of an already instantiated scene.\nTo create a variation of a scene, you can make an inherited scene based on the instantiated scene using Scene > New Inherited Scene... instead."));
+					accept->popup_centered();
+					continue;
```
This means that in case of multiple errors, the dialog will repeatedly set text and popup. While in the end it works, I wonder if there is a better way to do it 🤔

Xenthera replied:
Do you have any suggestions in mind? My thought is to show a popup that says "The following nodes were not copied" with the node names, but that wouldn't give a specific reason why. Another idea is just appended each node and its error to a string and display potentially a large amount of text in the popup. That doesn't sound like the best solution though either. Open to suggestions

KoBeWi replied:
Maybe some sort of bitmask/bool set for errors and if error appears, mark it in the set. Then after the loop you could check which error was flagged, add them to the dialog and display it all.
The affected scenes could be listed simply, but with a limit to e.g. 5 entries. So in the end the message would be:
```
These nodes failed to be saved: Node1, Node2, Node3, Node4, Node5 and X more nodes.
- Can't save the branch of an already instantiated scene.\nTo create a variation of a scene, you can make an inherited scene based on the instantiated scene using Scene > New Inherited Scene... instead.
- Etc.
```
Perfectly each failed node could be listed next to fail reason, but since the errors are so long, it's unfeasible for a simple dialog and improving it is out of scope now I'd say.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -974,60 +972,69 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
-				accept->popup_centered();
-				break;
-			}
+				List<String> extensions;
+				Ref<PackedScene> sd = memnew(PackedScene);
+				ResourceSaver::get_recognized_extensions(sd, &extensions);
```
This should be outside the loop.

Xenthera committed:
Merge branch 'godotengine:master' into master

Xenthera committed:
PR Feedback

Xenthera commented on editor/scene_tree_dock.cpp:
```diff
@@ -305,7 +305,6 @@ void SceneTreeDock::_replace_with_branch_scene(const String &p_file, Node *base)
 	instantiated_scene->set_unique_name_in_owner(base->is_unique_name_in_owner());
 
 	EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
-	undo_redo->create_action(TTR("Replace with Branch Scene"));
```
Added a check to only create/commit one undo action through either code flow. Thanks for pointing that out.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -1024,17 +1029,19 @@ void SceneTreeDock::_tool_selected(int p_tool, bool p_confirm_override) {
 					new_scene_from_dialog->popup_file_dialog();
 				} else {
-					if (da->file_exists(current_dir + existing)) {
+					String dir = new_scene_from_dialog->get_current_dir().path_join(existing);
```
It would be better to store `new_scene_from_dialog->get_current_dir()` in a variable outside the loop, to not call it for each node.

Xenthera committed:
Merge https://github.com/godotengine/godot into godotengine-master

# Conflicts:
#	editor/filesystem_dock.cpp

Xenthera committed:
Merge branch 'godotengine-master'
